File: editors/vscode/test/tilemap.test.js
'use strict';
// The map editor must read and draw a map as the engine does. testdata/tilemap (the farm
// the engine's own tilemap tests use) is loaded and drawn here in JavaScript and compared,
// pixel by pixel, with the goldens the engine writes (tilemap TestPicture, TestEdgeAtlas);
// the checks are compared with asset/map.go's messages.

const test = require('node:test');
const assert = require('node:assert');
const fs = require('fs');
const path = require('path');
const tex = require('../lib/texture');
const tm = require('../lib/tilemap');
const png = require('../lib/png');

const root = path.resolve(__dirname, '..', '..', '..');
const dir = path.join(root, 'testdata', 'tilemap');
const goldenDir = path.join(root, 'testdata', 'golden');
const haveEngine = fs.existsSync(dir);
const skip = haveEngine ? false : 'the engine\'s testdata is not here';

// library compiles the farm's textures as the editor does (tilemap_test.go testLibrary).
function library() {
  const textures = {};
  for (const name of ['grass', 'water', 'sand', 'dirt']) {
    const out = tex.compile(tex.parse(fs.readFileSync(path.join(dir, name + '.vtex'), 'utf8')).src, {});
    assert.deepStrictEqual(out.errors, [], name);
    textures[name] = tm.textureOf(name, out);
  }
  return { textures, materials: {} };
}

function farm() {
  const out = tm.load(fs.readFileSync(path.join(dir, 'farm.vmap'), 'utf8'), 'farm.vmap');
  assert.deepStrictEqual(out.errors, []);
  return out.map;
}

// same compares two RGBA pictures and names the first pixel that differs.
function same(got, file) {
  const want = png.decode(fs.readFileSync(path.join(goldenDir, file)));
  assert.strictEqual(`${got.w}x${got.h}`, `${want.w}x${want.h}`);
  for (let i = 0; i < want.data.length; i++) {
    if (want.data[i] !== got.data[i]) {
      const p = i >> 2;
      assert.fail(`${file}: channel ${i % 4} of pixel (${p % want.w}, ${Math.floor(p / want.w)}) is ${got.data[i]}, want ${want.data[i]}`);
    }
  }
}

test('the edge atlas of the water is the engine\'s', { skip }, () => {
  const water = library().textures.water;
  const a = tm.edgeAtlas(water);
  assert.deepStrictEqual(a.grid, [1, 2], 'a column of the 2 frames');
  assert.strictEqual(a.play, 'flow');
  same(a, 'tilemap_edge_atlas.png');
});

test('the farm is drawn as the engine draws it', { skip }, () => {
  const lib = library();
  const m = farm();
  assert.deepStrictEqual(tm.cellSize(tm.prepare(m, lib)), [16, 16]);
  same(tm.picture(m, lib, 0, 20), 'tilemap_farm_picture.png');
  same(tm.picture(m, lib, 10, 20), 'tilemap_farm_picture_tick10.png');
});

test('drawing some cells again gives the whole picture', { skip }, () => {
  const lib = library();
  const m = farm();
  const p = new tm.Picture(m, lib);
  p.drawAll();
  tm.brush(m, 0, 5, 5, 3, 2);
  p.drawRect(3, 3, 7, 7);
  const whole = tm.picture(m, lib, 0, 20);
  assert.ok(Buffer.from(p.data).equals(Buffer.from(whole.data)));
  // The water moves at tick 10: only the cells by the water are drawn again.
  const changed = p.setTick(10, 20);
  assert.deepStrictEqual(changed, [1]);
  for (let y = 0; y < m.h; y++) {
    for (let x = 0; x < m.w; x++) {
      if (p.touches(x, y, changed)) {
        p.drawRect(x, y, x, y);
      }
    }
  }
  assert.ok(Buffer.from(p.data).equals(Buffer.from(tm.picture(m, lib, 10, 20).data)));
});

// cliffs adds the autotile cliffs (tilemap_test.go cliffsLibrary) to the library.
function cliffs() {
  const lib = library();
  const images = { 'cliff.png': png.decode(fs.readFileSync(path.join(dir, 'cliff.png'))) };
  const out = tex.compile(tex.parse(fs.readFileSync(path.join(dir, 'cliff.vtex'), 'utf8')).src, images);
  assert.deepStrictEqual(out.errors, []);
  assert.strictEqual(out.spec.autotile, true);
  lib.textures.cliff = tm.textureOf('cliff', out);
  const m = tm.load(fs.readFileSync(path.join(dir, 'cliffs.vmap'), 'utf8'), 'cliffs.vmap');
  assert.deepStrictEqual(m.errors, []);
  return { lib, m: m.map };
}

test('autotiles are drawn as the engine draws them', { skip }, () => {
  const { lib, m } = cliffs();
  assert.deepStrictEqual(tm.cellSize(tm.prepare(m, lib)), [16, 16]);
  same(tm.picture(m, lib, 0, 20), 'tilemap_cliffs_picture.png');
  same(tm.picture(m, lib, 10, 20), 'tilemap_cliffs_picture_tick10.png');
  // Painting redraws the cells around: an autotile cell looks at its neighbours.
  const p = new tm.Picture(m, lib);
  p.drawAll();
  tm.paint(m, 1, 6, 4, 2);
  p.drawRect(5, 3, 7, 5);
  assert.ok(Buffer.from(p.data).equals(Buffer.from(tm.picture(m, lib, 0, 20).data)));
});

test('autoPick', () => {
  assert.deepStrictEqual(tm.autoPick(255), { tiles: [7, 7, 7, 7], whole: true });
  assert.deepStrictEqual(tm.autoPick(0), { tiles: [0, 2, 12, 14], whole: false });
  assert.deepStrictEqual(tm.autoPick(255 & ~2), { tiles: [15, 15, 15, 15], whole: true }, 'only the NE corner missing: the lake\'s SW tile');
});

test('cellEdges', { skip }, () => {
  const lib = library();
  const m = farm();
  const draws = tm.prepare(m, lib);
  const at = (l, x, y) => {
    const out = [];
    tm.cellEdges(m, draws, l, x, y, (u, s, q) => out.push([m.terrains[u - 1].name, s, q]));
    return out;
  };
  // Grass at (2, 1): sand below it, and below-right; the water is further.
  assert.deepStrictEqual(at(0, 2, 1), [['sand', 0, 2], ['sand', 0, 3]]);
  // A lone water cell (16, 6) is a border in every quarter of the grass around it.
  assert.deepStrictEqual(at(0, 17, 7), [['water', 3, 0]], 'the corner by the diagonal');
  // An empty cell of the paths layer under a path.
  assert.deepStrictEqual(at(1, 10, 0).map((e) => e[0]), ['dirt', 'dirt']);
  assert.deepStrictEqual(at(0, 3, 4), [], 'water is the highest: nothing spills on it');
});

test('the checks report what asset/map.go reports', () => {
  const bad = (src, file = 'm.vmap') => tm.load(typeof src === 'string' ? src : JSON.stringify(src), file).errors.map(tm.describe);
  const base = () => ({ veduta: 'map/1', size: [3, 2], terrains: [{ key: '.', name: 'grass', texture: 'grass' }], layers: [{ name: 'ground', rows: ['...', '...'] }] });
  assert.deepStrictEqual(bad(base()), []);
  const cases = [
    [(m) => { m.size = [0, 2000]; }, ['size[0]: 0 out of range [1, 1024]', 'size[1]: 2000 out of range [1, 1024]',
      'layers[0].rows: 2 rows, want 1 (the map\'s size)', 'layers[0].rows[0]: 3 characters, want 1 (the map\'s size; a space is an empty cell)']],
    [(m) => { m.size = [3]; m.layers[0].rows = ['.']; }, ['size: want [columns, rows], got 1 numbers']],
    [(m) => { delete m.size; m.layers[0].rows = ['.']; }, ['size: is required ([columns, rows])']],
    [(m) => { m.tile = 0; }, ['tile: must be a positive number, got 0']],
    [(m) => { m.origin = [1, 2]; }, ['origin: want 3 numbers, got 2']],
    [(m) => { m.terrains = []; m.layers[0].rows = ['   ', '   ']; }, ['terrains: at least one terrain is required']],
    [(m) => { m.terrains.push({ key: '.', name: 'Grass', texture: 'x', material: 'y', tags: ['a', 'a', 'B'] }); }, [
      'terrains[1].key: duplicate key "." (first used by terrains[0])',
      'terrains[1].name: name "Grass" may only contain a-z, 0-9, \'_\' and \'-\' (not first)',
      'terrains[1].material: not allowed with texture: a terrain is drawn with one or the other',
      'terrains[1].tags[1]: duplicate "a"',
      'terrains[1].tags[2]: name "B" may only contain a-z, 0-9, \'_\' and \'-\' (not first)']],
    [(m) => { m.terrains.push({ key: ' ', name: 'grass' }); }, [
      'terrains[1].key: must be one printable ASCII character but space, \'"\' and \'\\\', got " "',
      'terrains[1].name: duplicate terrain name "grass" (first used by terrains[0])',
      'terrains[1]: needs a texture or a material']],
    [(m) => { m.terrains[0].key = 'ab'; }, ['terrains[0].key: must be one printable ASCII character but space, \'"\' and \'\\\', got "ab"']],
    [(m) => { m.layers[0].rows = ['...', '.x', '...']; m.layers[0].layer = 2000; m.layers[0].z = 2e6; }, [
      'layers[0].layer: 2000 out of range [-1000, 1000]',
      'layers[0].z: 2e+06 out of range [-1e+06, 1e+06]',
      'layers[0].rows: 3 rows, want 2 (the map\'s size)',
      'layers[0].rows[1]: 2 characters, want 3 (the map\'s size; a space is an empty cell)']],
    [(m) => { m.layers[0].rows = ['.x.', '..é']; }, [
      'layers[0].rows[0]: character "x" at column 1 is not a terrain\'s key (a space is an empty cell)',
      'layers[0].rows[1]: 4 characters, want 3 (the map\'s size; a space is an empty cell)']],
    [(m) => { m.layers.push({ name: 'ground', rows: ['   ', '   '] }); }, ['layers[1].name: duplicate layer name "ground" (first used by layers[0])']],
    [(m) => {
      m.objects = [{ name: 'a', at: [3, 0] }, { name: 'a', at: [1, 1], size: [3, 1], tags: ['x'], props: { 'a.b': 1, ok: null, n: 'x' } }, { name: 'c' }];
    }, [
      'objects[0].at: cell [3 0] is outside the 3 × 2 map',
      'objects[1].name: duplicate object name "a" (first used by objects[0])',
      'objects[1].size: [3 1] cells from [1 1] leave the 3 × 2 map (each at least 1)',
      'objects[1].props.a.b: a property name is 1-64 characters without spaces or \'.\', got "a.b"',
      'objects[1].props.ok: must be a string, a number or a boolean',
      'objects[2].at: is required ([column, row] of its top-left cell)']],
  ];
  for (const [edit, wants] of cases) {
    const m = base();
    edit(m);
    const got = bad(m).map((s) => s.replace(/^\d+:\d+: /, ''));
    assert.deepStrictEqual(got.filter((g) => wants.includes(g)), wants, got.join('\n'));
  }
  // Decoding: the first problem, located.
  assert.deepStrictEqual(bad('{\n  "veduta": "map/1",\n  "size": [3, 2.5]\n}'), ['3:15: size[1]: cannot use JSON number 2.5 as int']);
  assert.deepStrictEqual(bad('{"veduta": "map/1", "size": [3, 2], "glow": 1}'), ['1:37: glow: unknown field']);
  assert.deepStrictEqual(bad('{"veduta": "map/1", "terrains": [{"key": 1}]}'), ['1:42: terrains[0].key: cannot use JSON number as string']);
  assert.deepStrictEqual(bad('{"veduta": "map/1", "layers": {}}'), ['1:31: layers: cannot use JSON object as []asset.MapLayerSource']);
  assert.deepStrictEqual(bad('{"veduta": "map/1", "size": [1, 1], "size": [2, 2]}'), ['1:37: duplicate key "size"']);
  assert.deepStrictEqual(bad('{"veduta": "texture/1"}'), ['1:12: veduta: header is "texture/1", want "map/1"']);
  assert.deepStrictEqual(bad('{"size": [1, 1]}'), ['1:1: missing "veduta" header (want "map/1")']);
  assert.deepStrictEqual(bad('[1]'), ['1:1: source must be a JSON object']);
  assert.deepStrictEqual(bad('{\n  "veduta": "map/1",\n  "size": [1 1]\n}'), ['3:14: invalid JSON: invalid character \'1\' after array element']);
  assert.deepStrictEqual(bad('{"veduta": "map/1"'), ['1:19: unexpected end of JSON input']);
  assert.deepStrictEqual(bad(base(), 'Farm.vmap'), ['file name: map name "Farm" may only contain a-z, 0-9, \'_\' and \'-\' (not first)']);
  // An error is located at the value it names.
  const text = JSON.stringify(Object.assign(base(), { tile: -1 }), null, 2);
  const e = tm.load(text, 'm.vmap').errors[0];
  assert.strictEqual(text.split('\n')[e.line - 1].slice(e.col - 1), '"tile": -1'.slice(8));
});

test('the new map template is written as it is', () => {
  const file = path.join(root, 'template', 'new', 'map.vmap');
  if (!fs.existsSync(file)) {
    return;
  }
  const text = fs.readFileSync(file, 'utf8');
  const out = tm.load(text, 'map.vmap');
  assert.deepStrictEqual(out.errors, []);
  assert.strictEqual(tm.format(out.map), text);
});

test('the example of docs/map.md and the farm round-trip', { skip }, () => {
  const doc = fs.readFileSync(path.join(root, 'docs', 'map.md'), 'utf8');
  const example = /```json\n(\{\n  "veduta": "map\/1"[\s\S]*?)```/.exec(doc)[1];
  for (const text of [example, fs.readFileSync(path.join(dir, 'farm.vmap'), 'utf8')]) {
    const a = tm.load(text, 'farm.vmap');
    assert.deepStrictEqual(a.errors, []);
    const once = tm.format(a.map);
    const b = tm.load(once, 'farm.vmap');
    assert.deepStrictEqual(b.errors, []);
    assert.strictEqual(tm.format(b.map), once, 'formatting is stable');
    assert.deepStrictEqual(JSON.parse(once), JSON.parse(text), 'the same map');
  }
  const once = tm.format(tm.load(example, 'farm.vmap').map);
  assert.ok(once.includes('    { "key": "s", "name": "sand", "texture": "sand" },\n'), 'a terrain per line');
  assert.ok(once.includes('    { "name": "paths", "rows": [\n      "        d   ",\n'), 'a row per line');
  assert.ok(once.includes('    { "name": "house_door", "at": [8, 0], "tags": ["door"], "props": { "to": "house", "x": 3, "y": 7 } },\n'));
});

test('cell operations', () => {
  const m = tm.load(JSON.stringify({ veduta: 'map/1', size: [5, 4], terrains: [{ key: '.', name: 'a', texture: 'a' }, { key: '~', name: 'b', texture: 'b' }],
    layers: [{ name: 'g', rows: ['.....', '.~~..', '.~...', '.....'] }, { name: 'p', z: 2, layer: 1, rows: ['     ', '     ', '     ', '     '] }],
    objects: [{ name: 'door', at: [4, 3] }, { name: 'field', at: [1, 1], size: [3, 2], props: { k: true } }] }), 'm.vmap').map;
  const show = (l) => tm.rows(m, l).join('|');
  assert.deepStrictEqual(tm.brush(m, 1, 0, 0, 2, 1), [[0, 0], [1, 0], [0, 1], [1, 1]]);
  assert.strictEqual(show(1), '..   |..   |     |     ');
  assert.strictEqual(tm.brush(m, 1, 2, 2, 3, 2).length, 9);
  assert.strictEqual(show(1), '..   |.~~~ | ~~~ | ~~~ ');
  assert.strictEqual(tm.paint(m, 1, 9, 9, 1).length, 0, 'off the map');
  assert.strictEqual(tm.rect(m, 1, 4, 3, 3, 2, 0).length, 2);
  assert.strictEqual(show(1), '..   |.~~~ | ~~  | ~~  ');
  // The fill stays on its layer and follows sides, not corners.
  assert.strictEqual(tm.flood(m, 0, 1, 1, 1).length, 3);
  assert.strictEqual(show(0), '.....|.....|.....|.....');
  assert.strictEqual(tm.flood(m, 0, 0, 0, 2).length, 20);
  assert.strictEqual(tm.flood(m, 0, 0, 0, 2).length, 0, 'same terrain: nothing');
  // Resize keeps the top-left, empties new cells, cuts objects and drops those off it.
  assert.deepStrictEqual(tm.resize(m, 3, 5), ['door']);
  assert.strictEqual(show(0), '~~~|~~~|~~~|~~~|   ');
  assert.deepStrictEqual(m.objects.map((o) => [o.name, o.size]), [['field', [2, 2]]]);
  const text = tm.format(m);
  assert.deepStrictEqual(tm.load(text, 'm.vmap').errors, []);
  assert.ok(text.includes('    { "name": "p", "z": 2, "layer": 1, "rows": [\n'));
  assert.ok(text.includes('"props": { "k": true }'));
  // Terrains: added with a free name and key, removed with their cells.
  assert.strictEqual(tm.addTerrain(m, 'a'), 2);
  assert.deepStrictEqual([m.terrains[2].name, m.terrains[2].key], ['a_2', '#']);
  assert.strictEqual(tm.uses(m, 1), 12 + 6);
  tm.paint(m, 0, 0, 4, 3);
  tm.removeTerrain(m, 1);
  assert.strictEqual(show(0), '   |   |   |   |#  ');
  assert.deepStrictEqual(tm.load(tm.format(m), 'm.vmap').errors, []);
  assert.strictEqual(new Set(tm.KEYS).size, tm.MaxMapTerrains);
});

File: tilemap/tilemap_test.go
package tilemap

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/riftbane/veduta/v2/asset"
	"github.com/riftbane/veduta/v2/asset/texture"
	"github.com/riftbane/veduta/v2/gfx"
	"github.com/riftbane/veduta/v2/gfx/soft"
	"github.com/riftbane/veduta/v2/gmath"
	"github.com/riftbane/veduta/v2/internal/golden"
	"github.com/riftbane/veduta/v2/scene"
)

// testLibrary compiles testdata/tilemap: the textures of a farm (grass, sand and water with
// borders, water animated, a mark in the grass's top-left texel to show which way up it is
// drawn) and the farm's map. The VS Code extension's tests read the same files.
func testLibrary(t testing.TB) *asset.Library {
	t.Helper()
	dir := filepath.Join("..", "testdata", "tilemap")
	lib := asset.NewLibrary(nil)
	for _, name := range []string{"grass", "water", "sand", "dirt"} {
		src, err := os.ReadFile(filepath.Join(dir, name+".vtex"))
		if err != nil {
			t.Fatal(err)
		}
		tx, err := texture.Parse(name+".vtex", src, texture.Options{})
		if err != nil {
			t.Fatal(err)
		}
		lib.Textures[name] = tx
	}
	src, err := os.ReadFile(filepath.Join(dir, "farm.vmap"))
	if err != nil {
		t.Fatal(err)
	}
	m, err := asset.ParseMap("farm.vmap", src)
	if err != nil {
		t.Fatal(err)
	}
	lib.Maps["farm"] = m
	return lib
}

// render draws the map with an orthographic camera looking at (x, y) showing height
// meters, at tick.
func render(t *testing.T, m *Map, lib *asset.Library, x, y, height float32, w, h int, tick uint64) *gfx.Image {
	t.Helper()
	r := soft.New(soft.Options{})
	defer r.Close()
	textures := map[string]*asset.Texture{}
	for k, v := range lib.Textures {
		textures[k] = v
	}
	for k, v := range m.Textures() {
		textures[k] = v
	}
	models := m.Models()
	res, err := scene.Upload(r, models, textures, m.Materials())
	if err != nil {
		t.Fatal(err)
	}
	s := scene.New("test", res.Bounds)
	s.Background = 0xff101018
	var dl gfx.DrawList
	s.Draw(&dl, res, scene.DrawOptions{Camera: scene.Camera2D(gmath.V2(x, y), height), Width: w, Height: h,
		Tick: tick, Rate: 20, Statics: m.Statics()})
	fb := gfx.NewFramebuffer(w, h, false)
	if err := r.Begin(fb); err != nil {
		t.Fatal(err)
	}
	if err := r.Draw(&dl); err != nil {
		t.Fatal(err)
	}
	if err := r.End(); err != nil {
		t.Fatal(err)
	}
	return fb.Image()
}

// TestRender: cells, borders by priority (water over sand over grass, a path layer's
// border over the ground), the corners of lone cells and the ripple of animated water.
func TestRender(t *testing.T) {
	lib := testLibrary(t)
	m := New(lib.Maps["farm"], lib)
	golden.Image(t, "tilemap_farm", render(t, m, lib, 0, 0, 18, 400, 360, 0))
	golden.Image(t, "tilemap_farm_tick10", render(t, m, lib, -5, 4, 6, 320, 240, 10))
}

// TestPicture: the map composed texel by texel, as the inspector and the VS Code map editor
// show it, at the start and when the water has moved to its second frame.
func TestPicture(t *testing.T) {
	lib := testLibrary(t)
	m := New(lib.Maps["farm"], lib)
	if w, h := m.CellSize(); w != 16 || h != 16 {
		t.Fatalf("cell %d × %d", w, h)
	}
	golden.Image(t, "tilemap_farm_picture", m.Picture(0, 20))
	golden.Image(t, "tilemap_farm_picture_tick10", m.Picture(10, 20))
}

func TestEdgeAtlas(t *testing.T) {
	lib := testLibrary(t)
	a := EdgeAtlas("map:edge:water", lib.Textures["water"])
	if a.Grid != [2]int{1, 2} || a.Play != "flow" || a.Clip("flow") == nil {
		t.Fatalf("atlas grid %v, play %q: want a column of the water's 2 frames and its clips", a.Grid, a.Play)
	}
	if img := a.Data.Levels[0]; img.W != 4*10 || img.H != 2*4*10 {
		t.Fatalf("atlas %d × %d, want 40 × 80: 4 × 4 quarters of 8 × 8 pixels padded to 10, per frame", img.W, img.H)
	}
	golden.Image(t, "tilemap_edge_atlas", a.Data.Levels[0])
}

func TestCells(t *testing.T) {
	lib := testLibrary(t)
	m := New(lib.Maps["farm"], lib)
	if w, h := m.Size(); w != 20 || h != 18 {
		t.Fatalf("size %d × %d", w, h)
	}
	// The origin is the top-left corner: cell (0, 0) spans x -10..-9, y 9..8.
	for _, c := range []struct {
		x, y   float32
		cx, cy int
	}{{-10, 9, 0, 0}, {-9.5, 8.5, 0, 0}, {-9, 8.99, 1, 0}, {-10.01, 9, -1, 0}, {0, 0, 10, 9}, {9.99, -8.99, 19, 17}} {
		if cx, cy := m.CellAt(c.x, c.y); cx != c.cx || cy != c.cy {
			t.Errorf("cell at (%v, %v) = (%d, %d), want (%d, %d)", c.x, c.y, cx, cy, c.cx, c.cy)
		}
	}
	if x, y := m.Center(0, 0); x != -9.5 || y != 8.5 {
		t.Errorf("center of (0, 0) = (%v, %v)", x, y)
	}
	if g := m.Get(0, 3, 4); g == nil || g.Name != "water" {
		t.Errorf("get (3, 4) = %v, want water", g)
	}
	if m.Get(1, 0, 0) != nil || m.Get(0, -1, 0) != nil || m.Get(2, 0, 0) != nil {
		t.Errorf("empty, off-map and missing-layer cells are not nil")
	}
	if !m.Has(-1, 10, 1, "path") || m.Has(0, 10, 1, "path") || !m.Has(0, 10, 1, "tillable") || m.Has(-1, 0, 0, "water") {
		t.Errorf("has: tags by layer wrong")
	}
	var events []string
	m.OnEvent = func(name string, f map[string]any) {
		events = append(events, name+" "+f["layer"].(string)+" "+f["terrain"].(string))
	}
	if err := m.Set(0, 0, 0, "water"); err != nil || m.Get(0, 0, 0).Name != "water" {
		t.Fatalf("set: %v", err)
	}
	if err := m.Set(1, 10, 1, ""); err != nil || m.Get(1, 10, 1) != nil {
		t.Fatalf("clear: %v", err)
	}
	m.Set(1, 10, 2, "dirt") // unchanged: no event
	if strings.Join(events, "; ") != "map_set ground water; map_set paths " {
		t.Errorf("events %q", events)
	}
	for _, c := range []struct {
		l, x, y int
		terrain string
		want    string
	}{{0, 20, 0, "water", "off the 20 × 18 map"}, {3, 0, 0, "water", "no layer 3"}, {0, 0, 0, "lava", `no terrain "lava" (it has grass, water, sand, dirt)`}} {
		if err := m.Set(c.l, c.x, c.y, c.terrain); err == nil || !strings.Contains(err.Error(), c.want) {
			t.Errorf("set %+v: %v, want %q", c, err, c.want)
		}
	}
	if err := m.SetCells([][]uint8{{1}}); err == nil {
		t.Errorf("SetCells with one layer accepted")
	}
}

// TestRebuild: painting a cell rebuilds the chunks whose look it changes, the chunk it is in
// and those its border reaches, and no other.
func TestRebuild(t *testing.T) {
	lib := testLibrary(t)
	m := New(lib.Maps["farm"], lib)
	before := map[string]*asset.Model{}
	for k, v := range m.Models() {
		before[k] = v
	}
	if len(before) != 7 { // 2 × 2 chunks, 2 layers, but no path near the bottom right
		t.Fatalf("%d chunk models, want 7", len(before))
	}
	for _, c := range []struct {
		x, y int
		want []string
	}{
		{5, 5, []string{"map:farm:0:0:0"}},                    // well inside chunk (0, 0)
		{15, 5, []string{"map:farm:0:0:0", "map:farm:0:1:0"}}, // its border reaches chunk (1, 0)
	} {
		m.Set(0, c.x, c.y, "dirt")
		after := m.Models()
		for name, md := range after {
			changed := md != before[name]
			if want := strings.Contains(strings.Join(c.want, " "), name); changed != want {
				t.Errorf("painting (%d, %d): %s rebuilt: %v, want %v", c.x, c.y, name, changed, want)
			}
			before[name] = md
		}
	}
	// Emptying every cell of a chunk, and those whose border reaches it, drops its model.
	for y := 15; y < 18; y++ {
		for x := 0; x < 16; x++ {
			m.Set(1, x, y, "")
		}
	}
	if _, ok := m.Models()["map:farm:1:0:1"]; ok {
		t.Errorf("an empty chunk still has a model")
	}
	if n := len(m.Statics()); n != 6 {
		t.Errorf("%d statics, want 6", n)
	}
}

// cliffsLibrary adds testdata/tilemap's cliffs to testLibrary: an autotile (from a PNG, its
// second frame lighter) over grass, in islands, lakes, lines, lone cells and cells that
// touch at a corner or the map's side.
func cliffsLibrary(t testing.TB) *asset.Library {
	t.Helper()
	lib := testLibrary(t)
	dir := filepath.Join("..", "testdata", "tilemap")
	src, err := os.ReadFile(filepath.Join(dir, "cliff.vtex"))
	if err != nil {
		t.Fatal(err)
	}
	tx, err := texture.Parse("cliff.vtex", src, texture.Options{FS: os.DirFS(dir)})
	if err != nil {
		t.Fatal(err)
	}
	lib.Textures["cliff"] = tx
	if src, err = os.ReadFile(filepath.Join(dir, "cliffs.vmap")); err != nil {
		t.Fatal(err)
	}
	m, err := asset.ParseMap("cliffs.vmap", src)
	if err != nil {
		t.Fatal(err)
	}
	lib.Maps["cliffs"] = m
	return lib
}

// TestAutotileClasses: the tile each class of quarter is taken from has that class there,
// and every tile but the lake's middle is drawn whole by the cell that looks like it.
func TestAutotileClasses(t *testing.T) {
	for c, tiles := range classTile {
		for q, tile := range tiles {
			if got := tileClasses[tile][q]; got != c {
				t.Errorf("class %d quarter %d: tile %d has class %d there", c, q, tile, got)
			}
		}
	}
	seen := map[[quarters]int]int{}
	for tile, cls := range tileClasses {
		if tile == lakeEmpty {
			continue
		}
		if other, ok := seen[cls]; ok {
			t.Errorf("tiles %d and %d have the same classes %v", other, tile, cls)
		}
		seen[cls] = tile
	}
	if tiles, whole := autoPick(0xff); !whole || tiles[0] != 7 {
		t.Errorf("a cell among its own: %v %v, want the island's middle whole", tiles, whole)
	}
	if tiles, whole := autoPick(0); whole || tiles != [quarters]int{0, 2, 12, 14} {
		t.Errorf("a lone cell: %v %v, want the island's four corners", tiles, whole)
	}
}

// TestAutotile: cliffs drawn by the renderer and composed by Picture, and the atlas.
func TestAutotile(t *testing.T) {
	lib := cliffsLibrary(t)
	m := New(lib.Maps["cliffs"], lib)
	if w, h := m.CellSize(); w != 16 || h != 16 {
		t.Fatalf("cell %d × %d, want a tile", w, h)
	}
	a := m.Textures()["map:auto:cliff"]
	if a == nil || a.Grid != [2]int{1, 2} || a.Play != "shine" {
		t.Fatalf("atlas %+v: want a column of the 2 frames and the clips", a)
	}
	if img := a.Data.Levels[0]; img.W != 6*18 || img.H != 2*3*18 {
		t.Fatalf("atlas %d × %d, want 108 × 108: 6 × 3 tiles padded to 18, per frame", img.W, img.H)
	}
	golden.Image(t, "tilemap_auto_atlas", a.Data.Levels[0])
	golden.Image(t, "tilemap_cliffs", render(t, m, lib, 0, 0, 12, 320, 240, 0))
	golden.Image(t, "tilemap_cliffs_picture", m.Picture(0, 20))
	golden.Image(t, "tilemap_cliffs_picture_tick10", m.Picture(10, 20))
}
